tests/virt_autotest/sriov_pci_passthrough.pm

```diff
@@ -0,0 +1,381 @@
+      return 1;
+    }
+
+    #check again the remaining vf inside vm
+    test_network_interface($guest, gate => $gateway, mac => $vf_vm_2{mac}, net => 'br123');
+    set_var("SRIOV_PCI_PASSSHTROUGH", 0);   #turn off to avoid affecting other tests
+
+    #detach the remaining vf from vm via virsh
+    record_info("Unplug VF from vm", "$host_vf_2{id} \nGuest: $guest \nbdf='$vf_vm_2{bdf}'   mac='$vf_vm_2{mac}'");
+    if (unplug_vf_from_vm($guest, $vf_vm_2{mac}, $vf_vm_2{xml})) {
+      $self->{test_results}->{$guest}->{"Error: fail to detach the first SR-IOV Ethernet device from $guest!"}->{status} = 'FAILED';       
+      return 1;
+    }
+  
+    #reattach vfs to host
+    record_info("Reattach VF to host", "vm=$guest \nvf=$host_vf_1{id} \nvf=$host_vf_2{id}");
+    assert_script_run("virsh nodedev-reattach $host_vf_1{id}", 60);
+    assert_script_run("virsh nodedev-reattach $host_vf_2{id}", 60);
+    script_run "lspci | grep Ethernet";
+  
+    #check host and guest to make sure they work well
+    if (check_host_and_guest($self, $guest)) {   #have not figured the parameters out???
+      $self->{test_results}->{$guest}->{"Error: host or guest $guest does not work well after vfs hot unplug from guest!"}->{status} = 'FAILED';
+    }
+
+  } #end each
+
+}
+
+
+#set up ssh, packages and iommu on host
+sub prepare_host(){
+
+  #generate ssh key on host
+  script_run "rm ~/.ssh/id_rsa*; ssh-keygen -t rsa -f ~/.ssh/id_rsa -P ''";
+
+  #install required packages on host 
+  zypper_call '-t in pciutils nmap';  #to run 'lspci' and 'nmap' ommand
+
+  #check IOMMU on XEN is enabled
+  if (get_var("XEN") || check_var("SYSTEM_ROLE", "xen") || check_var("HOST_HYPERVISOR", "xen")) {
+    assert_script_run "xl dmesg | grep IOMMU | grep -i enabled";
+  }
+}
+
+
+#get the BDF the PF device on host
+sub get_sriov_ethernet_devices () {
+
+  #get the BDF of the ethernet devices with SR-IOV
+  my $nic_devices = script_output "lspci | grep Ethernet | grep -v 'Virtual Function' | cut -d ' ' -f1" ;
+  my @nic_devices = split ("\n", $nic_devices);
+  my @sriov_devices;
+  foreach (@nic_devices){
+    if (! script_run "lspci -v -s $_ | grep -q 'SR-IOV'") {
+      push @sriov_devices, $_;
+    }
+  }
+  record_soft_failure "Error: there is no ethernet devices with SR-IOV in the host and test aborts!" if ! $#sriov_devices;
+  
+  return @sriov_devices;
+}
+
+#enable 8 virtual functions for the specified physical functions of the SR-IOV network device 
+sub enable_vf(){
+  my @pfs = @_;
+  
+  # get the network device drivers
+  my (@drivers, $pf, $driver);
+  foreach $pf (@pfs) {
+    $driver = script_output "lspci -v -s $pf | sed -n '/Kernel modules/p' | sed 's/.*Kernel modules: *//'";
+    my $is_known_driver = 0;
+    foreach (@drivers) {
+      next if ($_ ne "$driver");
+      $is_known_driver = 1;
+    }
+    push @drivers, $driver if ! $is_known_driver;
+  }
+  foreach (@drivers) {
+    assert_script_run("modprobe $driver", 60);
+  }
```

Comment:
If only one instruction inside suggest:
assert_script_run xxxx  foreach (@dirvers); 
And the current assert_script_run does not iterate over @drivers

Reply:
Merged to another each loop

Reply:
Merged to another each loop